devnet-sdk/testing/systest/multi_client.go

```diff
@@ -186,15 +186,54 @@ func (mc *MultiClient) fetchWithConsistencyCheck(
 		return nil, err
 	}
 
+	// Get the state root for the primary
+	primaryHeader, ok := primaryItem.(*types.Header)
+	if !ok {
+		// If we're fetching a block instead of a header
+		if block, ok := primaryItem.(*types.Block); ok {
+			primaryHeader = block.Header()
+		}
+	}
+
 	// Create a hash-only getter for followers
 	getFollowerHash := func(client HeaderProvider, num *big.Int) (common.Hash, error) {
-		_, _, hash, err := queryFn(client, num)
-		return hash, err
+		item, _, hash, err := queryFn(client, num)
+		if err != nil {
+			return common.Hash{}, err
+		}
+
+		// If hashes don't match, also check state roots
+		if hash != primaryHash && primaryHeader != nil {
+			var followerHeader *types.Header
+			if header, ok := item.(*types.Header); ok {
+				followerHeader = header
+			} else if block, ok := item.(*types.Block); ok {
+				followerHeader = block.Header()
+			}
+
+			// If we have both headers and the state roots match despite hash mismatch,
+			// include this information in the error message
+			if followerHeader != nil {
+				if followerHeader.Root == primaryHeader.Root {
+					return hash, fmt.Errorf("block hash divergence with matching state roots: primary_hash=%s, follower_hash=%s, state_root=%s",
+						primaryHash.Hex(), hash.Hex(), primaryHeader.Root.Hex())
+				} else {
+					// Both hash and state root differ
+					return hash, nil
```

Comment:
Shouldn't we return an error here if things aren't matching?

Reply:
Yes, we should. Good catch.